jep/0000/README.adoc

```diff
@@ -0,0 +1,283 @@
+//
+//
+// Uncomment and fill if this JEP is rendered obsolete by a later JEP
+//| Superseded-By
+//| :bulb: JEP-NUMBER :bulb:
+//
+//
+// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
+//| Resolution
+//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:
+
+|===
+
+
+== Abstract
+
+[TIP]
+====
+Give a short (200 word) description of the technical issue addressed.
+====
+
+Jenkins does not provide a homogeneous, easy to understand and easy use external API
+(through CLI and HTTP) to manage Jenkins data.
+
+Main contributions to this problem are: **lack of a real Resources API to be exposed** (raw model objects and sometimes
+just human readable text is returned nowadays), and **outdated and limited tooling** (CLI and HTTP APIs serialization
+is based on Stapler and Stapler annotations).
+
+This JEP is introducing both a new set of APIs to guide developers to write better external APIs and bringing
+new tooling to easily define the Resources layer to be exposed, as well as modern ways to build an HTTP/REST API.
+
+== Specification
+
+[TIP]
+====
+Provide a detailed specification what is being proposed.
+Be as technical and detailed as needed to allow new or existing Jenkins developers
+to reasonably understand the scope/impact of an implementation.
+====
+
+Main goals:
+
+. Isolate Model from public API: introduce a new _Resources layer_ based on new APIs (in advance the _Data API_)
+that can be leveraged for any data exposure (CLI and WS initially). It brings flexibility and improved
+API UX for CLI/WS users and developers.
+. Model can evolve whilst keeping backward compatibility in the Resources layer and Resources can
+transform, map or simplify the real model to adapt it for consumers.
+. The new _Data API_ is decoupled from Jenkins CLI and/or REST endpoints: the way to expose data is not a responsibility of the data itself.
+(However, this JEP also includes a proposal to expose data coming from the Data API through an HTTP API)
+. API messages are versioned (note the "version" attribute in messages) for easier evolution and correct handling of breaking changes.
+. Modernize Jenkins: using https://github.com/FasterXML/jackson[Jackson] instead of Stapler to define
+and run serialization and http://sparkjava.com/[Spark] to define/expose the HTTP API.
+. Why _Data API_? Because this JEP focuses on data access (Create, Read, Update and Delete of any Jenkins stored model
+using homogeneous access regardless the public interface used for it).
+. The _Data API_ will provide a mechanism to generate and parse JSON messages using `@Symbol` from the structs-plugin, so messages can look like:
+```
+{
+    "id": "github-key",
+    "type": "sshPrivateKey"
+    "scope": "GLOBAL",
+    "description": "GitHub checkout key",
+    "username": "john.doe",
+    "privateKey": "<redacted>",
+    "passphrase": "<redacted>"
+}
+```
+Note the `type` attribute, which is defined in the resource class as:
```

Comment:
I think that it may be worth trying to implement some form of hierarchy/namespacing in your proposal. At the moment, I imagine a flat sea of different classes that are all serialized with `@Symbol` which could raise two potential problems:
1) It becomes less clear to a user debugging the system (or reading HTTP responses) _where_ a particular data message (like the sshPrivateKey above) is defined, especially if they have many closely-related model that they would like to distinguish from other ones
2) You have the potential for two different plugins to define the same `@Symbol` name, leading to a naming clash and the data-api layer not being able to determine which concrete class to deserialise the data to.

Reply:
> It becomes less clear to a user debugging the system

That's not a user, it's probably a developer?

> You have the potential for two different plugins to define the same `@Symbol`

Yeah, I realized this could be a problem when the implementation of this comes... however I think the same problem exists in Pipeline. Need to check that.

Reply:
Naming convention/uniqueness was something that came to my mind as well.  Not sure yet what might be the best approach to mitigate that though.